### DISCUSSION TITLE
re-init displays

### DISCUSSION THREAD

**Discussion by Adrianotiger** [lovyan03/LovyanGFX#370] [Q&A]
Hi,
I have some ST7789 display connected to an ESP32S3 board (GPIO 35, 36, 37, 38).

I would like to init them all at the same time and it works (I have the control of the CS):
```cpp
void initDisplay()
{
  for(uint8_t j=0;j<4;j++)
  {
    digitalWrite(_CS[j], LOW);
  }
    // reset SPI (we need at least a 10us pulse to send a reset)
  delay(1);
  digitalWrite(PIN_RST, LOW); 
  delay(1); 
  digitalWrite(PIN_RST, HIGH); 

  // ST7789
  // Chip need at least 5ms to reset. 120ms when in sleep mode.
  delay(120);

  _tft->init();
  _tft->setSwapBytes(true);
  //_tft->setRotation(0);

  // ChipSelect can be disabled
  for(uint8_t j=0;j<4;j++)
  {
    digitalWrite(_CS[j], HIGH);
  }
}
```
And it works. 

I don't know why, after some hours a display can stop working. So I was thinking to re-init the display.
I added this code at the beginning of the `initDisplay()` function:

```cpp
_tft->getPanel()->getBus()->release();
delay(10);
_tft->getPanel()->getBus()->init();
delay(10);
```
Now, when I want to draw a text on one of the 4 displays, the display on GPIO37 is writing it too! I don't know if it is related with my ESP32-S3. How can I release the bus and re-init the display driver again?

Or is there a way to detect if a ST7789 is not working like expected without MISO (since my display doesn't have it)?

Thanks.

**Comment by tobozo** [comment]
hi, thanks for your feedback :+1: 

> (GPIO 35, 36, 37, 38).

Are those are the respective CS pins for each panel?
I'm assuming all four panels have their own LGFX object+config, and the configuration blocks only differ in their `pin_cs` value.

> after some hours a display can stop working

This isn't the first time I've heard about a secondary panel doing strange stuff after a while with S3 boards, can you provide more information about the display loop, maybe share some code where the problem is reproductible?

> So I was thinking to re-init the display.

This would only patch the symptoms, and leave the root cause in the unknown. 

>  is there a way to detect if a ST7789 is not working like expected

Without understanding the cause it's difficult to know what to look for, but a few blind tests can be done.


### check if tft is readable

```cpp
uint16_t test_color = tft->color565( 0x12, 0x34, 0x56);
tft->fillRect( x, y, 1, 1, test_color ); // write a pixel
if( tft->readPixel( x, y ) != test_color ) { // read the pixel from TFT
   // fillRect() and/or readPixel() failed !
}
```

### Using write transaction separately for each display.

```cpp

LGFX* tft[4] = {&panel1, &panel2, &panel3, &panel4};

void loop()
{
    for(int i=0;i<4;i++ ) {
      tft[i]->startWrite()
      drawPanel( i ); // draw your stuff
      tft[i]->endWrite();
      tft[i]->display();
      tft[i]->waitDisplay();
    }

}

```

This will only tell if the problem is happening on the bus though.



### Use different CS Pin

I know all pins are suppposed to act alike, but if GPIO37 is acting strangely it could be interesting to observe what happens when the CS line of this panel is assigned to another GPIO.

**Comment by Adrianotiger** [comment]
Thank you very much for the fast response Tobozo!

I does not create a new object for each display. I am using just one object and setting the chipSelect when I want to update a specific display. In this way, I can send an image on all displays at the same time for example.

I am now able to read the display pixel and see if the display is not responding anymore. But how can I re-init the display?

Here a test code (now with 7 displays :P):

```cpp
#include <SPI.h>
#define LGFX_USE_V1
#include <LovyanGFX.hpp>

const uint8_t NUM_DISPLAYS = 7;
const uint8_t PIN_RST = 14;
const uint8_t PIN_DC = 9;
const uint16_t TFT_W = 240;
const uint16_t TFT_H = 240;

uint8_t _CS[NUM_DISPLAYS] = {41,40,39,38,37,36,35}; // ChipSelect
uint8_t _BL[NUM_DISPLAYS] = { 8,16,15, 7, 6, 5, 4}; // BackLight
uint8_t _TL[NUM_DISPLAYS] = {10, 1, 2,42,48,47,21}; // Test Led

class LGFX : public lgfx::LGFX_Device
{
  lgfx::Panel_ST7789    _panel_instance;
  lgfx::Bus_SPI         _bus_instance;
public:
  LGFX(void) {
    {
      auto cfg = _bus_instance.config();
      cfg.spi_host    = SPI2_HOST;
      cfg.spi_mode    = 3;
      cfg.freq_write  = 1000000;
      cfg.freq_read   = 1000000;
      cfg.spi_3wire   = true;
      cfg.use_lock    = false;
      cfg.dma_channel = SPI_DMA_CH_AUTO;
      cfg.pin_sclk    = SCK;
      cfg.pin_mosi    = MOSI;
      cfg.pin_miso    = -1;    // -1=disable
      cfg.pin_dc      =  PIN_DC;
      _bus_instance.config(cfg);
      _panel_instance.setBus(&_bus_instance);
    }

    {
      auto cfg = _panel_instance.config();
      cfg.pin_cs           =    -1;
      cfg.pin_rst          =    -1;
      cfg.pin_busy         =    -1;
      cfg.panel_width      =   TFT_W;
      cfg.panel_height     =   TFT_H;
      cfg.memory_width     =   TFT_W;
      cfg.memory_height    =   TFT_H;
      cfg.offset_x         =     0;  
      cfg.offset_y         =     0;  
      cfg.offset_rotation  =     0;  
      cfg.dummy_read_pixel =     8;  
      cfg.dummy_read_bits  =     1;  
      cfg.readable         =  true;  
      cfg.invert           =  true;  
      cfg.rgb_order        = false;  
      cfg.dlen_16bit       = false;  
      cfg.bus_shared       = false;

      _panel_instance.config(cfg);
    }
    setPanel(&_panel_instance);
  }
};

LGFX *_tft;

void initBus()
{
  /* Init SPI */
  for(uint8_t j=0;j<NUM_DISPLAYS;j++) digitalWrite(_CS[j], LOW);

    // reset SPI devices
  delay(1);
  digitalWrite(PIN_RST, LOW); 
  delay(1); 
  digitalWrite(PIN_RST, HIGH); 
  delay(20);

  _tft->init();
  _tft->setSwapBytes(true);
  log_v("Lovyan TFT initiated");

  for(uint8_t j=0;j<NUM_DISPLAYS;j++) digitalWrite(_CS[j], HIGH);
}

void setup() 
{
  Serial.begin(115200);
  for(uint8_t j=0;j<NUM_DISPLAYS;j++) 
  {
    pinMode(_CS[j], OUTPUT);
    pinMode(_BL[j], OUTPUT);
    pinMode(_TL[j], OUTPUT);

    digitalWrite(_BL[j], HIGH); // power on backlight
    digitalWrite(_TL[j], LOW);  // power off testled
    digitalWrite(_CS[j], HIGH);  // inactive chip select
  }

  // set common SPI outputs
  pinMode(PIN_DC, OUTPUT);
  pinMode(PIN_RST, OUTPUT);
  digitalWrite(PIN_DC, HIGH);
  digitalWrite(PIN_RST, HIGH); 

  _tft = new LGFX();

  delay(500);

  initBus();

  delay(100);
  // Upgrade SPI Speed from 1MHz to 10MHz (just a test)
  auto _panel_instance = _tft->getPanel();
  auto _bus_instance = (lgfx::Bus_SPI*)_panel_instance->getBus();
  auto _bus_cfg = _bus_instance->config();
  _bus_cfg.freq_write = 10000000L;
  _bus_instance->config(_bus_cfg);
  _panel_instance->setBus( _bus_instance );

}

void loop() 
{
  uint16_t color = 0xffff;
  // reset bus 
  /*
   * If released and re-init, the display on GPIO37 has a strange behaviour (like draw when CS is not active) 
  */
  /*
  Serial.println("Resetting bus...");
  _tft->getPanel()->getBus()->release();
  delay(10);
  _tft->getPanel()->getBus()->init();
  delay(10);
  initBus();
  delay(100);
  */

  for(uint8_t k=0;k<11*NUM_DISPLAYS;k++)
  {
      // blink first test led
    digitalWrite(_TL[0], (k%2)==0 ? HIGH : LOW);

    if((k % 11) == 0) // read pixel)
    {
      digitalWrite(_CS[0], LOW);
      uint16_t pc = _tft->readPixel( 100, 100 );
      Serial.print("Pixel color: 0x");
      Serial.println(pc, HEX);
      digitalWrite(_CS[0], HIGH);
    }

    for(uint8_t j=0;j<NUM_DISPLAYS;j++)
    {
      digitalWrite(_CS[j], LOW);
      _tft->fillScreen(color);
      digitalWrite(_CS[j], HIGH);
    } 
    color = (color < 0x0008) ? 0xffff : color >> 0x05;
  }
}

```

If I call `tft->init()` a second time, the displays will not work anymore.
I can call ` _tft->getPanel()->getBus()->release()` and then ` _tft->getPanel()->getBus()->init();` before init the displays again. But with this procedure, the display on GPIO37 loose the CS and so it will show every image sent to the other displays.

Is there a better way to reset or just re-init a single display, when I detect that it doesn't update the display anymore? It seems that the library does not like a second `init()`.

**Comment by tobozo** [comment]
yup multiple tft.init() is a bit messy

can you try this in your loop() ?

```cpp

    for(uint8_t j=0;j<NUM_DISPLAYS;j++)
    {
      digitalWrite(_CS[j], LOW);
      _tft->->startWrite()
      _tft->fillScreen(color);
      _tft->->endWrite();
      _tft->->display();
      _tft->->waitDisplay();
      digitalWrite(_CS[j], HIGH);
    } 

```

**Comment by Adrianotiger** [comment]
Hi Tobozo.
If I insert it in the code, I can't see any difference.

> yup multiple tft.init() is a bit messy  

I was never able to connect multiple displays using multiple objects. I don't know if it up to me or this library just doesn't like to be instanced more than 1 time.
For example, I can't connect 4 displays and instantiate the same class 4 times. Even if I create the object with a class:
```
LGFX *_tft1 = new LGFX(cs1);
LGFX *_tft2 = new LGFX(cs2);
LGFX *_tft3 = new LGFX(cs3);
LGFX *_tft4 = new LGFX(cs4);
```
It seems that there are some interferences internally (with the bus itself or with the reset line?).
But I can create 1 LGFX instance and manage the chip selects manually.

I know, this is a specific case with a specific display driver. But I believe it is always possible that a chip will stop responding and you need to reset it. Or not?

I updated the code, increasing the frequency from 1MHz up to 80MHz (with a 1MHz step). At ~40MHz the first display begins to answer with the wrong pixel color.
Code:
```cpp
uint8_t errors = 0;
for(uint8_t j=0;j<NUM_DISPLAYS;j++)
{
  digitalWrite(_CS[j], LOW);
  uint16_t pixelColor = _tft->readPixel( 100, 100 );
  digitalWrite(_CS[j], HIGH);
  Serial.print("Display"); Serial.print(j+1, DEC);
  Serial.print("- color: 0x");  Serial.println(pc, HEX);
  if(pixelColor != color) 
    errors++;
}
```

I can then check the status of the display with this code:
```cpp
lgfx::Panel_ST7789 *_panel_instance = (lgfx::Panel_ST7789*)_tft->getPanel();
...
if(errors > 0)
{
  for(uint32_t j=0;j<NUM_DISPLAYS;j++)
  {
    digitalWrite(_CS[j], LOW);
    uint32_t panelStatus= _panel_instance->readCommand(0x09, 0, 4);
    digitalWrite(_CS[j], HIGH);
    Serial.print("Display "); Serial.print(j, DEC);
    Serial.print("status: 0x"); Serial.println(panelStatus, HEX);
  }
}
```

I am here now, and I see that the status of 1 display is different from the others on the same bus. But I am still looking for a solution when the display (or ST7789) doesn't answer anymore.

Maybe with a SW-reset command (0x01) to the display.
Or setting the SPI-speed to 1MHz.

My first display is 5cm away and my last display is 50cm away from this ESP32. This doesn't help :P

Thank you again for your help! I was able to continue debugging with your help!

**Reply by tobozo** [reply]
a bit hacky but  you can try tft->sleep() and  tft.wakeup() when an error is detected with readPixel test

**Comment by Adrianotiger** [comment]
Ok, it was partially my fault.
During my tests, I tried to increase the clock up to 80MHz. But the display-driver (st7789) can't go over 62.5MHz.

At this point the display, changed the registry settings:
- sometimes the colors were inverted
- sometimes the rotation was changed
- sometimes I just didn't saw anything on the screen anymore

I tried with sleep and wakeup. But it doesn't restore the tft "health" with this sequence.
If I resend the command_list of the display panel after a SW-RESET, I can restore and make it works again :P 

Or I just stay in the range, so I will never get this strange effect, hehe

Maybe adding a parameter to the configuration, like "max_freq_write" and "max_freq_read" to the panel devices, could prevent this problem?
Example: 
on the ST7789 the `max_freq_write` would be 62.5MHz and `max_freq_read` would be 6.6MHz.

**Comment by lovyan03** [comment]
Just to be sure, is there an SD card (TF card) connected on the same SPI bus?
If an SD card is present and you have not gone through the procedure to change the SD to SPI mode, the SD card may malfunction and prevent communication with the display.

**Comment by Adrianotiger** [comment]
Hi Lovyan san,
no there isn't any SPI device connected. Only displays (all the same type).
Edit: I am also not using the touch or other functionality of this library. Only the ST7789 are connected on the bus.
